### PR TITLE
update CRI(LinuxContainerResources ) to support hugepages

### PR DESCRIPTION
**Pull request by bg-chun**
Signed-off-by: Byonggon Chun <bg.chun@samsung.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This CRI update is intended to support:
1) [container isolation of huge pages](https://github.com/kubernetes/enhancements/pull/1199)
2) [multi size hugepages at container level](https://github.com/kubernetes/enhancements/pull/1271)

**Which issue(s) this PR fixes**:
https://github.com/kubernetes/kubernetes/issues/80716
https://github.com/kubernetes/kubernetes/issues/77251

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP]: [Update huge pages KEP for container isolation of huge pages](https://github.com/kubernetes/enhancements/pull/1199)
- [KEP]: [Huge pages KEP: support multiple sizes huge pages on a container level](https://github.com/kubernetes/enhancements/pull/1271)
- [google docs]: [Hugepages in 1.17](https://docs.google.com/document/d/1Xnb3ioEH_JmWLdFJ-i6Lp7jFc_fDuzzAa2oEXJglBbM/edit#)
- [google slide]: [Why container isolation instead of pod isolation
](https://docs.google.com/presentation/d/1H5xhCjgCZJjdTm-mGUFwqoz8XmoVd8xTCwJ1AfVR_54/edit#slide=id.g6384351b13_0_3)

### CODE REVIEW

**Review by mattjmcnaughton — COMMENTED** [review]
cc @odinuge 

/ok-to-test

Adding some additional reviewers who have offered insights on huge pages in the past :)

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto
Comment (kad) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: should it be `uint` that can be matched with corresponding entry in `/sys/devices/system/node/node*/hugepages/hugepages-*kB`  ?

Comment (bg-chun) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: `Or idea to validate it always with what is available in /sys/fs/cgroup/hugetlb/hugetlb.* and fail if something doesn't match?`
=> Yes, it is my original thought.
I referred OCI [runtime spec](https://github.com/opencontainers/runtime-spec/blob/19e92ca817772b4466f2ed2b8d808dfb7a8ab4be/specs-go/config.go#L220).
Shoud i update it to `<unit-prefix-in-capital-letter>` to avoid ambiguity?
```
// LinuxHugepageLimit structure corresponds to limiting kernel hugepages
type LinuxHugepageLimit struct {
	// Pagesize is the hugepage size
	// Format: "<size><unit-prefix>B' (e.g. 64KB, 2MB, 1GB, etc.)
	Pagesize string `json:"pageSize"`
	// Limit is the limit of "hugepagesize" hugetlb usage
	Limit uint64 `json:"limit"`
}
```

**Comment by bg-chun** [comment]
Adding more people for reviews.
/cc @bartosh
/cc @sjenning 
/cc @PiotrProkop

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto
Comment (kad) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: I think it is ok, but updating comment referencing OCI spec https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#huge-page-limits can be also a good idea. OCI spec is explicit here to define what this string should contain.

Comment (odinuge) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: I think updating the comment is nice too.

I agree with @kad that using an uint would be nice, but I still thing following the OSI spec is the best solution, so I prefer to keep it as a string. The size prefixes originate from here btw: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/mm/hugetlb_cgroup.c?id=e3280b54afed870d531571212f1fc375df39b7d2#n342

**Review by odinuge — APPROVED** [review]
/lgtm

**Review by odinuge — APPROVED** [review]
/lgtm

**Comment by derekwaynecarr** [comment]
I would like to hold on this PR until we have the associated KEP merged.

If we can update the KEP to eliminate the discussion around topology alignment and just keep it focused on adding hugepage information to LinuxContainerResources, it would help focus.

**Comment by bg-chun** [comment]
@derekwaynecarr 
Can we consider this CRI update for 1.17?
I don't mean that "let's merge every single PRs of hugepage for 1.17".
I`m not sure that we can make `container isolation of hugepage` for 1.17.
But I think we can merge it for 1.18 and update of container runtimes.

Container runtimes have to been updated to support the CRI change(this PR).
For example, `containerd` needs to update `cri-plugin`;
and docker requires more expensive changes(`dockershim`, and etc).

**Comment by derekwaynecarr** [comment]
should container stats also report usage?

**Comment by bg-chun** [comment]
@derekwaynecarr 
As I know current kubectl doesn't shows stats of hugepages.
I think it will be a little bit helpful if kubectl shows hugepages stats.
But I don't have use case for it.
And as I know, node stats report hugepages usage.

**Comment by derekwaynecarr** [comment]
I am fine adding per container stats later.

@yujuhong i will hold this for a couple days to give your acl, otherwise will just merge.

/hold

**Comment by derekwaynecarr** [comment]
@mrunalp this will be needed for crio.

**Comment by mrunalp** [comment]
CRI-O PR open here - https://github.com/cri-o/cri-o/pull/2940

**Comment by bg-chun** [comment]
@yujuhong 
https://docs.google.com/presentation/d/1J_6RGeNU1f21TchLrE672SfFLcbhPUev356NLGYxoUU/edit#slide=id.g7a7421ffe7_0_50
On this slide, there is my overall plan.
I will try add e2e test and update cri-tools  for validation.

**Comment by yujuhong** [comment]
> https://docs.google.com/presentation/d/1J_6RGeNU1f21TchLrE672SfFLcbhPUev356NLGYxoUU/edit#slide=id.g7a7421ffe7_0_50

Thanks for the write-up!

**Pull request merged**
